src/Aws/Kinesis/Client/Consumer.hs

```diff
@@ -241,93 +220,92 @@ managedKinesisConsumer kit =
 -- it.
 --
 withKinesisConsumer
-  ∷ MonadConsumer m
+  ∷ ( MonadIO m
+    , MonadBaseControl IO m
+    )
   ⇒ ConsumerKit
   → (KinesisConsumer → m α)
   → m α
-withKinesisConsumer kit inner =
-  flip runReaderT kit $ do
-    batchSize ← view ckBatchSize
-    messageQueue ← liftIO ∘ newTBQueueIO $ batchSize * 10
-
-    state ← updateStreamState CR.empty ≫= liftIO ∘ newTVarIO
-    let reshardingLoop = forever $
-          handleError (\_ → liftIO $ threadDelay 3000000) $ do
-            liftIO (readTVarIO state)
-              ≫= updateStreamState
-              ≫= liftIO ∘ atomically ∘ writeTVar state
-            liftIO $ threadDelay 10000000
-
-        producerLoop = forever $
-          handleError (\_ → liftIO $ threadDelay 2000000) $ do
-            recordsCount ← replenishMessages messageQueue state
-            liftIO ∘ threadDelay $
-              case recordsCount of
-                0 → 5000000
-                _ → 70000
-
-
-    withAsync reshardingLoop $ \reshardingHandle → do
-      link reshardingHandle
-      withAsync producerLoop $ \producerHandle → do
-        link producerHandle
-        res ← lift ∘ inner $ KinesisConsumer messageQueue state
-        return res
+withKinesisConsumer kit inner = do
+  let batchSize = kit ^. ckBatchSize
+  messageQueue ← liftIO ∘ newTBQueueIO $ batchSize * 10
+
+  state ← liftIO $ updateStreamState kit CR.empty ≫= newTVarIO
+
+  let
+    reshardingLoop =
+      forever ∘ handle (\(SomeException _) → threadDelay 3000000) $ do
+        readTVarIO state
+          ≫= updateStreamState kit
+          ≫= atomically ∘ writeTVar state
+        threadDelay 10000000
+
+    producerLoop =
+      forever ∘ handle (\(SomeException _) → threadDelay 2000000) $ do
+        recordsCount ← replenishMessages kit messageQueue state
+        threadDelay $
+          case recordsCount of
+            0 → 5000000
+            _ → 70000
```

Comment:
Right; very sadly, these constants are the result of many weeks of tweaking in order to avoid rate limiting errors. 

- [ ] I'll add a comment to that effect.